magic_the_gathering/game_state.py

```diff
@@ -1,53 +1,61 @@
 
 from magic_the_gathering.cards.base import Card
-
-
-class CardPosition(Enum):
-    UNTAPPED = 0
-    TAPPED = 1
-
-
-@dataclass
-class BoardItem:
-    card: Card
-    state: CardPosition
-
-
-@dataclass
-class StackItem:
-    card: Card
-    owner_player_index: int
+from magic_the_gathering.game_modes.base import GameMode
+from magic_the_gathering.players.base import Player
 
 
 class GameState:
     def __init__(
         self,
+        game_mode: GameMode,
+        players: List[Player],
+        current_turn_counter: Optional[int] = 0,
+        current_player_index: Optional[int] = 0,
+        current_player_has_played_a_land_this_turn: Optional[bool] = False,
         hands: Optional[List[List[Card]]] = None,
         decks: Optional[List[List[Card]]] = None,
-        boards: Optional[List[List[BoardItem]]] = None,
-        stack: Optional[List[StackItem]] = None,
-        life_totals: Optional[List[int]] = None,
+        boards: Optional[List[List[Card]]] = None,
+        graveyards: Optional[List[List[Card]]] = None,
+        exiles: Optional[List[Card]] = None,
+        stack: Optional[List[Card]] = None,
     ):
+        self.game_mode = game_mode
+        self.players = players
+        self.current_turn_counter = current_turn_counter
+        self.current_player_index = current_player_index
+        # TODO: Should 'current_player_has_played_a_land_this_turn' be a property of the player, not of the game state?
+        # TODO: At some point we'll change this so that we allow a player to have n land drops, 1 by default, but can change with some cards
+        self.current_player_has_played_a_land_this_turn = current_player_has_played_a_land_this_turn
         self.hands = hands
         self.decks = decks
         self.boards = boards
+        self.graveyards = graveyards
+        self.exiles = exiles
         self.stack = stack
-        self.life_totals = life_totals
 
-    def initialize_for_new_game(self, initial_life_points, n_players=2):
+    @property
+    def n_players(self):
+        return len(self.players)
+
+    @property
+    def current_player(self):
+        return self.players[self.current_player_index]
+
+    def initialize_for_new_game(self):
+        # TODO: It assumes that the players were initialized, is it a good idea?
+        # FIXME: The game mode contains information about the player initial life points and hand size, but it is not used here
+        # TODO: Pass the current player index? (for now it is always 0)
         # Set hands, decks, boards and stack to empty lists
-        self.hands = [[] for _ in range(n_players)]
-        self.decks = [[] for _ in range(n_players)]
-        self.boards = [[] for _ in range(n_players)]
+        self.hands = [[] for _ in range(self.n_players)]
+        self.decks = [[] for _ in range(self.n_players)]
+        self.boards = [[] for _ in range(self.n_players)]
+        self.graveyards = [[] for _ in range(self.n_players)]
+        # TODO: Is there one exile per player?
```

Comment:
Yes there is one exile zone per player, but there are complications.
There are cards in exile that can be linked to another card, like the card exiled by gonti https://scryfall.com/card/clb/753/gonti-lord-of-luxury or with the adventure mechanic https://scryfall.com/card/clb/781/bonecrusher-giant-stomp
Also, there are some additional exile-like zones like the companion zone or commander zone, but I doubt we want to worry about that for now

Reply:
Ok thanks. Regarding the companion and commander zones, I think we can exclude them for now. Regarding the link between cards, could we have a class which ties two cards together? If I understood well, when we equip an artifact to a creature, a link is created as well (although it might be of a different kind, since for instance when the creature dies the artifact stays, which is not the case for enchantment if I remember well). Anyway, my point is that linking cards seem to be a very common mechanic in Magic so we could add classes to support it :slightly_smiling_face: